x/wasm/internal/keeper/recurse_test.go

```diff
@@ -243,3 +224,88 @@ func TestGasOnExternalQuery(t *testing.T) {
 		})
 	}
 }
+
+func TestLimitRecursiveQueryGas(t *testing.T) {
+	// The point of this test from https://github.com/CosmWasm/cosmwasm/issues/456
+	// Basically, if I burn 90% of gas in CPU loop, then query out (to my self)
+	// the sub-query will have all the original gas (minus the 40k instance charge)
+	// and can burn 90% and call a sub-contract again...
+	// This attack would allow us to use far more than the provided gas before
+	// eventually hitting an OutOfGas panic.
+
+	const (
+		// Note: about 100 SDK gas (10k wasmer gas) for each round of sha256
+		GasWork2k uint64 = InstanceCost + 233_379 // we have 6x gas used in cpu than in the instance
+		// This is overhead for calling into a sub-contract
+		GasReturnHashed uint64 = 603
+	)
+
+	cases := map[string]struct {
+		gasLimit       uint64
+		msg            Recurse
+		expectedGas    uint64
+		expectOutOfGas bool
+	}{
+		"no recursion, lots of work": {
+			gasLimit: 4_000_000,
+			msg: Recurse{
+				Depth: 0,
+				Work:  2000,
+			},
+			expectedGas: GasWork2k,
+		},
+		"recursion 5, lots of work": {
+			gasLimit: 4_000_000,
+			msg: Recurse{
+				Depth: 5,
+				Work:  2000,
+			},
+			expectedGas: GasWork2k + 5*(GasWork2k+GasReturnHashed),
+		},
+		// this is where we expect an error...
+		// it has enough gas to run 4 times and die on the 5th
+		// however, if we don't charge the cpu gas before sub-dispatching, we can recurse over 20 times
+		// TODO: figure out how to asset how deep it went
```

Comment:
This is what needs to be tested... any ideas how?

Once I get this failing (more than 5 levels), then I can work on fixing the go-cosmwasm levels and proving this worked.

Reply:
🤔  You could modify the `Querier` with a custom `QueryPlugin` and increment a counter on each call so that you would know how often it was called.

Reply:
That's an interesting idea.
I was thinking to add some global stats counter to the keeper, but that seems cleaner if possible

Reply:
I started this approach here: https://github.com/CosmWasm/wasmd/pull/234 but it seems to require many deeper changes

Reply:
Oof, my idea also requires making it a pointer, seems like I have to suck that up and do it.